### ISSUE TITLE
Can we use previous Sensor Obs schemas and patterns

### ISSUE THREAD

**Issue by chris-little** [opengeospatial/Environmental-Data-Retrieval-API#59]
In the Sensor Web Enablement and SensorThings API groups and mailing lists, there was concern expressed that there was a lack of acknowledgement of previous OGC work covering data structures, metadata and queries relevant to EDR. In particular:
1. [Sensor Observation Services and its schemas](https://www.ogc.org/standards/sos) 
1. [WaterML](https://www.ogc.org/standards/waterml) 
1. and its direct derivative [TimeSeriesML](https://www.ogc.org/standards/tsml)  

SOS has XML schemas with filter patterns for time and space. Can we use these? If so, then we can add a reference to the standard.

**Comment by chris-little** [comment]
In the Sensor Web Enablement and SensorThings API groups and mailing lists, there was concern expressed that there was a lack of acknowledgement of previous OGC work covering data structures, metadata and queries relevant to EDR. In particular:
(Sensor Observation Services and its schemas)[https://www.ogc.org/standards/sos] 
(WaterML)[https://www.ogc.org/standards/waterml] 
and its direct derivative (TimeSeriesML)[https://www.ogc.org/standards/tsml]  
SOS has XML schemas with filter patterns for time and space. Can we use these? If so, then we can add a reference to the standard.

**Comment by dblodgett-usgs** [comment]
**`"lack of acknowledgement"` of what? is an important question.** 

Our aim here is to have a hypermedia-driven JSON-based API with (optionally) XML and other format data encodings. From that perspective, WaterML, TimeSeriesML, or SWE-Common-CSV are clearly "in-scope". I don't think we need to acknowledge specific payloads at this stage of development other than to say -- yes they are supported. (Note I have more on this front below the next point)

On the SOS-API front, **is there a SWE conceptual or logical schema that we could consider in our implementation?** We've been circling around and attempting to maintain consistency with the general O&M concepts (conceptual model?) in describing EDR as a "sampling API". Perhaps unpacking that for some readers concerned about carrying the SWE legacy forward would be worthwhile? 

Regarding interplay between payload encoding-content and API, I am extremely leery. Pushing the hypermedia too far down and coupling the API too tightly to any one encoding could paint us into a corner and risk ballooning the specification drastically. To take this point one step further, the complexity contained in legacy (XML) physical data models and the technical baggage they bring (I'm looking at you ... xlink) is exactly what we are trying to get away from.

So to sum up, I think **we should** absolutely continue to adhere rigorously to the concepts--and to a practical extent -- logical models of the SWE suite and SensorThings. I think we **should not** consider harmonizing or binding our conformance classes to them.

**Comment by ShaneMill1** [comment]
Somewhat related- in prototyping the search engine, I used an O&M example that contains swe. However, right now the search engine is utilizing csw and we obviously discussed using OGC-API Records and creating an extension for MetOcean needs that could include O&M and swe. Just mentioning this could be one area it is incorporated. 

Also, in mentioning TSML, that certainly could be an outputFormat for a specific query of data. In fact, that is what NWS is looking to do with NDFD data. However, in the scope of your question, that is just an optional outputFormat supplied from a provider. This ties directly to Dave's comment so I'm in agreement with him there.

**Comment by dr-shorthair** [comment]
I don't particularly care about _acknowledgement_ for its own sake. I also don't care about specific formats and payloads - that is essentially uninteresting (though if you get it wrong it can really bog you down - I'm looking at you metadata standards!). 

What I do care about is wheel re-invention. 

There is a body of work relating to services that deliver time-series and related datasets, which OGC has been standardizing since about 2003. Delivering TimeSeries from within datacubes is mostly what SOS does [1]. If EDR is a logical successor, then let us say that. If it is not, then what is the different problem that it solves? Related: the Sampling Features taxonomy in O&M v1 Part 2  [2] was derived from CSML, which was a product of the met-oceans community. If EDR is proposing a different, more efficient way of solving the same issue, then great - but lets be clear to the current users of SOS and the O&M Sampling Features structures (particularly in the INSPIRE community) that they should plan on coming over to EDR in due course. At the moment I see no signposts.  

What I am troubled by is the discontinuity in OGC initiatives that appear to be addressing the same problem. 

[1] https://www.ogc.org/standards/sos
[2] http://portal.opengeospatial.org/files/?artifact_id=22467 and its successor http://portal.opengeospatial.org/files/?artifact_id=41579 

In relation to Josh's comments on samples vs sampling etc, that was addressed further in the [OGC/W3C SSN ontology](https://www.w3.org/TR/vocab-ssn/), which has the notion of an 'act of sampling' to create a specific subset of the ultimate feature of interest, which could be a physical sample but could also be a trajectory or sounding etc.

**Comment by dblodgett-usgs** [comment]
Sentence 1 of the introduction of the SOS implementation spec is:

> The Sensor Observation Service (SOS) provides a standardized interface for managing and
retrieving metadata and observations from heterogeneous sensor systems.

While EDR could have some relation to that use case in some implementations of it, the technical use case we are working on here is a fundamentally different one. 

That said, we are hoping that this work is based on the same concepts and is ostensibly in-line with O&M, as it should be. However, the focus here is on adoptability in the modern web landscape; a bar that SOS does not meet.

If there are specific places where our work here, which has a distinct technical focus and use case, is out of line with the baseline of O&M-based implementation specs, we need those things called out and we need to address them in more specific issues than this. 

Regarding sign posts for users of SWE tech to use EDR instead, that's just not what we are aiming for here. There's no way this work is going to take on the breadth of sensor-based use cases that SWE does unless the SWE community things it would be worth adding a whole lot to a future EDR component.

@chris-little -- my vote is to answer no to your question and close this issue unless there are specific aspects of the identified standards that we should consider.

**Comment by dr-shorthair** [comment]
Oh dear. I really don't understand how 

> the technical use case we are working on here is a fundamentally different 

to 

> retrieving metadata and observations from heterogeneous sensor systems.

Note that the SOS GetObservation operation is parameterized by 
- featureOfInterest (i.e. location)
- observedProperty (i.e. variable or parameter)
- spatial and temporal filters (which could include the kind of sampling patterns you've been discussing)
- a few other things

See figure 9 on p29 in https://portal.opengeospatial.org/files/?artifact_id=47599. 

The response is usually dominated by a 'result' object, which is pretty much the kind of data structure that EDR talks about. 

Yeah, it is all wrapped up in some UML and XML, and some parts make it look like it is all about live sensor feeds, but in practice its almost always staged through a cache which could be a cube. So I don't understand how this is 'fundamentally different'. Arguably EDR is just a special case. That might allow for some simplification in implementation, but the concepts look pretty close to me. 

A related argument could be made as to how EDR is a special case of WCS. 

I'm not proposing that EDR is misconceived, just that its positioning as a successor interface to the earlier OGC technology is unclear, and an opportunity which should be noted. EDR is likely a _better_ solution for some existing users of older OGC tech. For example, the INSPIRE Environmental Monitoring facility spec [1] is cast in terms of O&M etc, but in practice is mostly about time-series data. 

[1] https://inspire.ec.europa.eu/Themes/120/2892

**Comment by dblodgett-usgs** [comment]
When I say a different technical use case, I mean that SOS is defining a physical data model focused on technical details of "heterogeneous sensor systems" that are totally different to what we are doing here:

> An Environmental Data Retrevial (EDR) API provides a family of lightweight interfaces to access Environmental Data resources. 

I think we may be talking past each other. What I'm saying is, since there is no conceptual model other than O&M, there isn't really anything for us to build on at the level we are working.

Looking at the specifics, 

![Screen Shot 2020-05-07 at 7 38 39 AM](https://user-images.githubusercontent.com/1492803/81295286-cfab5500-9035-11ea-962e-a2f6a3ffc32a.png)
[Figure 9 from OGC 12-0006](https://portal.opengeospatial.org/files/?artifact_id=47599)

Based on this figure, things that may be worth looking at are:
1) should we be teasing out procedure in the work we are doing? Currently, procedure is bound up in EDR dataset metadata and not exposed for query other than selection of a dataset and parameter.  
2) Should we use the observation offering pattern as a cataloging mechanism? Currently, we lump EDR datasets under collections to be consistent with the approach being pursued by other OGC API SWGs. There is ongoing debate around this practice that is a little more existential -- focused on having a common approach to generic "items" or allowing multiple datasets to be cataloged as collections in a single OGC API instance. The conceptual basis for this distinction so far is rooted in dataset distribution and NOT in observation offerings(collections). Personally, I think we should avoid the added complexity and focus on building blocks that could be used to create offerings/collections as the business of a given data provider or consortium of data providers prefers -- which would mean we would not use the offerings pattern at all.... but I digress.

Are there other details of the concepts here that you think we are out of sync with? My proposal is to close this issue as it is too vague and just saying we are reinventing wheels is unproductive. If the details I suggested just above or others people want to raise warrant additional issues be opened, let's do that.

**Comment by ShaneMill1** [comment]
> Somewhat related- in prototyping the search engine, I used an O&M example that contains swe. However, right now the search engine is utilizing csw and we obviously discussed using OGC-API Records and creating an extension for MetOcean needs that could include O&M and swe. Just mentioning this could be one area it is incorporated.
> 
> Also, in mentioning TSML, that certainly could be an outputFormat for a specific query of data. In fact, that is what NWS is looking to do with NDFD data. However, in the scope of your question, that is just an optional outputFormat supplied from a provider. This ties directly to Dave's comment so I'm in agreement with him there.

Just to follow up on this comment in case it was a bit misleading:
I don't think that this is within the scope of the actual specification. In terms of the search engine, that is an approach that could be a part of the Best Practices. In terms of TSML, that is an implementation choice for the provider, which also is not part of the specification except for outputFormat choices.

**Comment by chris-little** [comment]
@dr-shorthair I think that one difference between the SOS 
> "Delivering TimeSeries from within datacubes is mostly what SOS does"

 and the EDR is that the EDR data store is expected to be dense, and most queries within the geospatial and temporal extent of the store would return relevant data. My understanding that the SOS datacubes could be quite sparse, and 'finding where the data is' gives a different emphasis to solutions. So I do not think EDR is a replacement successor to SOS, but complementary.

**Comment by dr-shorthair** [comment]
@dblodgett-usgs my impression is that each EDR endpoint probably has a fixed value for `offering` and `procedure`. That is the sense in which I see EDR as a specialization of SOS. And of course it leans on more modern tech for implementing the interface. That is how it is a replacement successor for the most common use-case of SOS.

**Comment by dr-shorthair** [comment]
@chris-little created this issue in response to issues mentioned on email lists. In terms of the effect on the spec I suggest that a little additional context should be added in 

- 1.i - to explain where EDR fits into the OGC baseline conceptually - particularly in relation to SWE and Coverages/WCS
- 7.1 - a sentence at the end drawing attention to the fact that EDR functionality was supported in the W*S suite by SOS, and that this is a successor in the OGC API family 

Then perhaps we could recruit one of the SWE/SOS team to map a couple of the examples to SOS requests and put that in an appendix.

**Comment by dblodgett-usgs** [comment]
... so what you want is acknowledgement? *wink*

I think that sounds totally reasonable. Unless someone with a deeper background in the nuances of the specs wants to draft that text, I would be willing to take it on. Will table this issue till our next telecon.

**Comment by dr-shorthair** [comment]
No. (And I'm not an editor on any of those earlier specs anyway.) 

What I want is some signposts to users of the clumsy tools that they are currently using that their needs will be better met by this wondrous new product. And maybe a brief explanation of how EDR really is an implementation of concepts that have been in the OGC baseline for a long time. 

OGC had this pattern of Abstract Spec and Implementation Standard. At times it could get all a bit baroque and slow. But what you are doing here really is Coverages and Sensor data. So the models are already there. You are just building an efficient implementation using more modern tools.

**Comment by dblodgett-usgs** [comment]
Totally understood and appreciated.

**Comment by chris-little** [comment]
@dblodgett-usgs Your offer to draft sentences to add the spec noted!
